src/v/coproc/script_context_backend.cc

```diff
@@ -64,37 +76,96 @@ static ss::future<bool> do_write_materialized_partition(
     co_await std::move(compressed)
       .for_each_ref(log.make_appender(write_cfg), model::no_timeout)
       .discard_result();
-    co_return true;
 }
 
 static ss::future<storage::log>
 get_log(storage::log_manager& log_mgr, const model::ntp& ntp) {
     auto found = log_mgr.get(ntp);
     if (found) {
+        /// Log exists, do nothing and return it
         return ss::make_ready_future<storage::log>(*found);
     }
-    vlog(coproclog.info, "Making new log: {}", ntp);
-    return log_mgr.manage(storage::ntp_config(ntp, log_mgr.config().base_dir));
+    /// In the case the storage::log has not been created, but the topic has,
+    /// register to recieve a notification when it is created.
+    ss::promise<storage::log> p;
+    auto f = p.get_future();
+    auto id = log_mgr.register_manage_notification(
+      ntp, [p = std::move(p)](storage::log log) mutable { p.set_value(log); });
+    return f.then([&log_mgr, id](storage::log log) {
+        log_mgr.unregister_manage_notification(id);
+        return log;
+    });
```

Comment:
This is concerning because (1) it alludes to a race and (2) it creates a lot of coupling with assumptions about how that race is solved. Generally such races are solved using retries.

It would be nice to remove the notification upcalls from the log manager. Can you make the storage::log a std::optional<storage::log> and then fix the race by trying to resolve the storage::log pointer when you encounter a nullopt and otherwise returning a retriable error?

Maybe to start you could explain a little bit about what the race is and why it needs to be solved this way?

Reply:
Agreed I wasn't too happy with this portion. I am happy to implement this with a retry. The root cause is that we cannot wait on this event to occur, since `create_materialized_topic` really only resolves when the topic table on all participants has had valid entries updated. Then out of band the `controller_backend` does the real work of updating the shard table and creating the `storage::log`, so you cannot wait on that event. I therefore created this mechanism to wait on the desired event instead of just retrying, thinking it would be cleaner, but admittedly there can be issue with this too as what you mentioned makes alot of sense.

Reply:
it doesn't look like this code is in the PR anymore, so do you want to resolve this and address in the next review?

Reply:
It is part of the requested change. I removed the changes that created and used a notification system within log manager, in favor of throwing a special exception which will invoke a retry